designs/conditional-controllers.md

```diff
@@ -0,0 +1,174 @@
+Informer Removal / Controller Lifecycle Management
+==========================
+
+## Summary
+
+Enable fine-grained control over the lifecycle of a controller, including the
+ability to start/stop/restart controllers and their caches by exposing a way to
+remove individual informers from the cache and working around restrictions that
+currently prevent controllers from starting multiple times.
+
+## Background/Motivation
+
+Currently, the user does not have much controller over the lifecycle of a
+controller. The user can add controllers to the manager and add informers to the
+cache, but there is no way to remove either of these.
+
+Additionally there are restrictions that prevent a controller from running
+multiple times such as the clearing the watches slice for a controller after it
+has started. The effect of this is that users have no clean way of restarting controllers
+after they have stopped them.
+
+Greater detail is given in the "Future Work / Use Cases" section, but a summary
+of the motivating use-cases for the proposal is anytime a controller's watched resource
+can be installed/uninstalled unexpectedly or whenever the administrator of a
+controller is different from the administrator of the cluster (and thus has no
+control over which resources are installed).
+
+## Goals
+
+An implementation of the minimally viable hooks needed in controller-runtime to
+enable controller adminstrators to start, stop, and restart controllers and their caches.
+
+## Non-Goals
+
+A complete and ergonomic solution for automatically starting/stopping/restarting
+controllers upon arbitrary conditions.
+
+## Proposal
+
+The following proposal offers a solution for controller/cache restarting by:
+1. Enabling the removal of individual informers.
+2. Publically exposing the informer removal and adding hooks into the internal
+   controller implementation to allow for restarting controllers that have been
+   stopped.
+
+This proposal focuses on solutions that are entirely contained in
+controller-runtime. In the alternatives section, we discuss potential ways that
+changes can be added to api-machinery code in core kubernetes to enable a
+possibly cleaner interface of accomplishing our goals.
+
+### Informer Removal
+
+The starting point for this proposal is Shomron’s proposed implementation of
```

Comment:
Please describe this a bit, rahter than just saying "This PR explains it", because the PR might change

Reply:
done